cluster-autoscaler/simulator/nodes.go

```diff
@@ -17,46 +17,43 @@ limitations under the License.
 package simulator
 
 import (
+	appsv1 "k8s.io/api/apps/v1"
 	apiv1 "k8s.io/api/core/v1"
+	"k8s.io/autoscaler/cluster-autoscaler/utils/daemonset"
+	"k8s.io/autoscaler/cluster-autoscaler/utils/errors"
 	schedulerframework "k8s.io/kubernetes/pkg/scheduler/framework"
 
-	"k8s.io/autoscaler/cluster-autoscaler/utils/errors"
 	pod_util "k8s.io/autoscaler/cluster-autoscaler/utils/pod"
 )
 
-// getRequiredPodsForNode returns a list of pods that would appear on the node if the
-// node was just created (like daemonset and manifest-run pods). It reuses kubectl
-// drain command to get the list.
-func getRequiredPodsForNode(nodename string, podsForNodes map[string][]*apiv1.Pod) ([]*apiv1.Pod, errors.AutoscalerError) {
-	allPods := podsForNodes[nodename]
-
-	return filterRequiredPodsForNode(allPods), nil
-}
-
 // BuildNodeInfoForNode build a NodeInfo structure for the given node as if the node was just created.
-func BuildNodeInfoForNode(node *apiv1.Node, podsForNodes map[string][]*apiv1.Pod) (*schedulerframework.NodeInfo, errors.AutoscalerError) {
-	requiredPods, err := getRequiredPodsForNode(node.Name, podsForNodes)
-	if err != nil {
-		return nil, err
-	}
-	result := schedulerframework.NewNodeInfo(requiredPods...)
-	result.SetNode(node)
-	return result, nil
+func BuildNodeInfoForNode(node *apiv1.Node, scheduledPods []*apiv1.Pod, daemonsets []*appsv1.DaemonSet, forceDaemonSets bool) (*schedulerframework.NodeInfo, errors.AutoscalerError) {
+	nodeInfo := schedulerframework.NewNodeInfo()
+	nodeInfo.SetNode(node)
+	return addExpectedPods(nodeInfo, scheduledPods, daemonsets, forceDaemonSets)
 }
 
-func filterRequiredPodsForNode(allPods []*apiv1.Pod) []*apiv1.Pod {
-	var selectedPods []*apiv1.Pod
-
-	for id, pod := range allPods {
-		// Ignore pod in deletion phase
+func addExpectedPods(nodeInfo *schedulerframework.NodeInfo, scheduledPods []*apiv1.Pod, daemonsets []*appsv1.DaemonSet, forceDaemonSets bool) (*schedulerframework.NodeInfo, errors.AutoscalerError) {
+	for _, pod := range scheduledPods {
+		// Ignore scheduled pods in deletion phase
 		if pod.DeletionTimestamp != nil {
 			continue
 		}
-
-		if pod_util.IsMirrorPod(pod) || pod_util.IsDaemonSetPod(pod) {
-			selectedPods = append(selectedPods, allPods[id])
+		// Add scheduled mirror pods and scheduled DS pods if not force scheduling DS
+		if pod_util.IsMirrorPod(pod) || (!forceDaemonSets && pod_util.IsDaemonSetPod(pod)) {
```

Comment:
This will ignore all DaemonSet-like (but not really DaemonSet) pods for which `IsDaemonSetPod` currently returns true. I think we need to distinguish between "real DS" and "DS-like" pods here to make this function work correctly.

Reply:
That is a good point. I think we can remedy this by checking which DS pods are already present on the chosen node and inserting new DS pods only for DSs that are not present. That way we will still keep all the "DS-like" pods to the template.

Reply:
Changed, PTAL